File: lib/services/calendar_schedule_generator.dart
import 'package:flutter/foundation.dart';

import '../core/date_utils.dart';
import '../core/template_origin.dart';
import '../data/db/drift_database.dart';
import '../features/calendar/domain/models.dart';
import 'user_profile_service.dart';
import 'package:wellness_app/services/language_service.dart';

class CalendarScheduleGenerator {
  final AppDatabase _database;
  final UserProfile _profile;

  /// The language the plan is being generated in.
  ///
  /// Event titles are *stored*, so this decides what a user sees on their
  /// calendar forever after. Both generators already fill `nameHe` on their
  /// templates; this class was copying `template.name` -- the English field --
  /// so a plan built during Hebrew onboarding came out in English even though
  /// the Hebrew names existed alongside it.
  final AppLanguage _language;

  CalendarScheduleGenerator(this._database, this._profile,
      [this._language = AppLanguage.english]);

  /// Picks the Hebrew name when generating in Hebrew and one exists.
  ///
  /// The *Data classes carry `nameHe` but not the `displayName()` the domain
  /// models have, so the same fallback rule lives here: an untranslated
  /// template keeps its English name rather than showing a blank.
  String _named(String name, String? nameHe) =>
      _language == AppLanguage.hebrew &&
              nameHe != null &&
              nameHe.trim().isNotEmpty
          ? nameHe
          : name;

  /// Builds the events a new user's calendar should start with.
  ///
  /// Returns them rather than saving: the caller persists them through
  /// `CalendarNotifier.addEvents`, which is the only path that also schedules
  /// notifications. This class used to call `CalendarService.saveEvent`
  /// directly, so four weeks of generated workouts produced zero reminders.
  ///
  /// Workouts, meals and sleep are all covered -- previously only workouts
  /// were generated, despite onboarding collecting a meal count and the
  /// settings screen offering meal and sleep reminders.
  Future<List<ScheduledEvent>> buildSchedule() async {
    return [
      ...await _buildWorkoutEvents(),
      ..._buildMealEvents(await _database.getAllMealTemplates()),
      _buildSleepEvent(),
    ];
  }

  Future<List<ScheduledEvent>> _buildWorkoutEvents() async {
    final templates = await _database.getAllWorkoutTemplates();

    // Prefer templates generated for *this* profile: those went through
    // ProfileFit, so every exercise in them is one the user owns equipment
    // for and isn't injured against. Built-ins are seeded before any profile
    // exists and may contradict it -- scheduling a barbell session for
    // someone with no equipment is exactly the mismatch this whole pass is
    // about. Fall back to built-ins only when nothing was generated.
    final generated =
        templates.where((t) => t.origin == TemplateOrigin.generated).toList();
    final mainTemplates = generated.isNotEmpty ? generated : templates;

    if (mainTemplates.isEmpty) {
      debugPrint(
          '[CALENDAR-GEN] No workout templates, skipping workout events');
      return [];
    }

    // One weekly-recurring event per training day, each pinned to a template.
    // Recurrence means the series keeps going instead of stopping after the
    // four weeks the old loop materialised.
    final events = <ScheduledEvent>[];
    final scheduleDays = _getScheduleDays();
    final firstDay = AppDateUtils.startOfDay(DateTime.now());

    for (var i = 0; i < scheduleDays.length; i++) {
      final weekday = scheduleDays[i];
      final template = mainTemplates[i % mainTemplates.length];
      final firstOccurrence = _nextWeekdayOnOrAfter(firstDay, weekday);

      events.add(ScheduledEvent.create(
        title: _named(template.name, template.nameHe),
        type: EventType.workout,
        scheduledAt: DateTime(firstOccurrence.year, firstOccurrence.month,
            firstOccurrence.day, _workoutHour),
        recurrenceType: RecurrenceType.weekly,
        recurrenceDays: [weekday],
        templateId: template.id,
        metadata: const {'leadMinutes': 10},
      ));
    }
    return events;
  }

  List<ScheduledEvent> _buildMealEvents(List<MealTemplateData> mealTemplates) {
    final times = _mealTimes();
    final events = <ScheduledEvent>[];
    final today = AppDateUtils.startOfDay(DateTime.now());

    for (var i = 0; i < times.length; i++) {
      final slot = times[i];
      // Pin to a generated template when there is one, so the notification's
      // "Approve" action has something to build the meal from.
      // Same preference as workouts: a generated template is guaranteed to
      // respect the user's diet and exclusions, a built-in is not.
      final generated = mealTemplates
          .where((t) => t.origin == TemplateOrigin.generated)
          .toList();
      final pool = generated.isNotEmpty ? generated : mealTemplates;
      final template = pool.isEmpty ? null : pool[i % pool.length];

      events.add(ScheduledEvent.create(
        title: template == null
            ? slot.label
            : _named(template.name, template.nameHe),
        type: EventType.meal,
        scheduledAt: DateTime(
            today.year, today.month, today.day, slot.hour, slot.minute),
        recurrenceType: RecurrenceType.daily,
        templateId: template?.id,
      ));
    }
    return events;
  }

  ScheduledEvent _buildSleepEvent() {
    final today = AppDateUtils.startOfDay(DateTime.now());
    return ScheduledEvent.create(
      title: _language == AppLanguage.hebrew ? 'שינה' : 'Sleep',
      type: EventType.sleep,
      scheduledAt: DateTime(today.year, today.month, today.day, _bedtimeHour),
      recurrenceType: RecurrenceType.daily,
    );
  }

  /// Meal slots for the profile's meal count. The labels are only a fallback
  /// for when no meal template was generated to name the event after, so they
  /// follow the generation language too.
  List<({String label, int hour, int minute})> _mealTimes() {
    switch (_profile.mealCountPerDay) {
      case '2':
        return [
          (
            label: _language == AppLanguage.hebrew ? 'ארוחת צהריים' : 'Lunch',
            hour: 12,
            minute: 30
          ),
          (
            label: _language == AppLanguage.hebrew ? 'ארוחת ערב' : 'Dinner',
            hour: 19,
            minute: 0
          ),
        ];
      case '4':
        return [
          (
            label: _language == AppLanguage.hebrew ? 'ארוחת בוקר' : 'Breakfast',
            hour: 8,
            minute: 0
          ),
          (
            label: _language == AppLanguage.hebrew ? 'ארוחת צהריים' : 'Lunch',
            hour: 12,
            minute: 30
          ),
          (
            label: _language == AppLanguage.hebrew ? 'חטיף' : 'Snack',
            hour: 16,
            minute: 0
          ),
          (
            label: _language == AppLanguage.hebrew ? 'ארוחת ערב' : 'Dinner',
            hour: 19,
            minute: 30
          ),
        ];
      case 'intermittent_fasting_16_8':
        // Eating window opens at midday.
        return [
          (
            label: _language == AppLanguage.hebrew ? 'ארוחת צהריים' : 'Lunch',
            hour: 12,
            minute: 0
          ),
          (
            label: _language == AppLanguage.hebrew ? 'חטיף' : 'Snack',
            hour: 16,
            minute: 0
          ),
          (
            label: _language == AppLanguage.hebrew ? 'ארוחת ערב' : 'Dinner',
            hour: 19,
            minute: 30
          ),
        ];
      case '3':
      default:
        return [
          (
            label: _language == AppLanguage.hebrew ? 'ארוחת בוקר' : 'Breakfast',
            hour: 8,
            minute: 0
          ),
          (
            label: _language == AppLanguage.hebrew ? 'ארוחת צהריים' : 'Lunch',
            hour: 12,
            minute: 30
          ),
          (
            label: _language == AppLanguage.hebrew ? 'ארוחת ערב' : 'Dinner',
            hour: 19,
            minute: 0
          ),
        ];
    }
  }

  static DateTime _nextWeekdayOnOrAfter(DateTime from, int weekday) {
    final delta = (weekday - from.weekday + 7) % 7;
    return from.add(Duration(days: delta));
  }

  static const int _workoutHour = 7;
  static const int _bedtimeHour = 22;

  /// Training days as `DateTime.weekday` numbers (1 = Monday .. 7 = Sunday).
  ///
  /// Filled Sunday through Thursday first, then Friday, then Saturday. The
  /// app follows the Israeli convention and treats Fri/Sat as the weekend
  /// everywhere else, so those are used only when the requested frequency
  /// leaves no choice -- which it does at 6 and 7 days.
  ///
  /// This used to cap at 5 no matter what was asked: someone who said "I
  /// train 6 days a week" silently got 5, and 0-1 days got 2. The answer was
  /// collected and then partly ignored, which is the same complaint ISSUES #7
  /// records against the template side.
  static const List<int> _weekdayFillOrder = [
    DateTime.sunday,
    DateTime.monday,
    DateTime.tuesday,
    DateTime.wednesday,
    DateTime.thursday,
    DateTime.friday,
    DateTime.saturday,
  ];

  List<int> _getScheduleDays() {
    final count = _profile.trainingDaysPerWeek.clamp(1, 7);

    // Spread the week evenly for the low frequencies rather than stacking
    // consecutive days: three sessions belong on Sun/Tue/Thu, not Sun/Mon/Tue.
    switch (count) {
      case 1:
        return [DateTime.sunday];
      case 2:
        return [DateTime.sunday, DateTime.wednesday];
      case 3:
        return [DateTime.sunday, DateTime.tuesday, DateTime.thursday];
      case 4:
        return [
          DateTime.sunday,
          DateTime.monday,
          DateTime.wednesday,
          DateTime.thursday,
        ];
      default:
        // 5, 6 and 7 take the fill order directly, so Friday and Saturday are
        // only ever reached once the working week is full.
        return _weekdayFillOrder.take(count).toList();
    }
  }
}

File: Packages/WellnessKit/Sources/WellnessDomain/CalendarScheduleGenerator.swift
import Foundation
import WellnessModels

public struct CalendarScheduleGenerator: Sendable {
    public let profile: UserProfile
    public let language: AppLanguage

    public init(profile: UserProfile, language: AppLanguage = .english) {
        self.profile = profile
        self.language = language
    }

    private func text(_ english: String, _ hebrew: String) -> String {
        language == .hebrew ? hebrew : english
    }

    public struct MealTemplateRef: Sendable {
        public let id: String
        public let name: String
        public let origin: TemplateOrigin

        public init(id: String, name: String, origin: TemplateOrigin) {
            self.id = id; self.name = name; self.origin = origin
        }
    }

    public struct WorkoutTemplateRef: Sendable {
        public let id: String
        public let name: String
        public let origin: TemplateOrigin

        public init(id: String, name: String, origin: TemplateOrigin) {
            self.id = id; self.name = name; self.origin = origin
        }
    }

    public func buildSchedule(
        workoutTemplates: [WorkoutTemplateRef],
        mealTemplates: [MealTemplateRef],
        now: Date = .now,
        makeId: () -> String
    ) -> [ScheduledEvent] {
        var events: [ScheduledEvent] = []
        events.append(contentsOf: buildWorkoutEvents(workoutTemplates, now: now, makeId: makeId))
        events.append(contentsOf: buildMealEvents(mealTemplates, now: now, makeId: makeId))
        events.append(buildSleepEvent(now: now, makeId: makeId))
        return events
    }

    // MARK: - Workouts

    private static let workoutHour = 7

    private func buildWorkoutEvents(
        _ templates: [WorkoutTemplateRef],
        now: Date,
        makeId: () -> String
    ) -> [ScheduledEvent] {
        let generated = templates.filter { $0.origin == .generated }
        let main = generated.isEmpty ? templates : generated
        guard !main.isEmpty else { return [] }

        let cal = Calendar(identifier: .gregorian)
        let firstDay = cal.startOfDay(for: now)
        let scheduleDays = getScheduleDays()
        var events: [ScheduledEvent] = []

        for (i, weekday) in scheduleDays.enumerated() {
            let template = main[i % main.count]
            let firstOccurrence = Self.nextWeekdayOnOrAfter(firstDay, weekday, cal: cal)
            var components = cal.dateComponents([.year, .month, .day], from: firstOccurrence)
            components.hour = Self.workoutHour

            events.append(ScheduledEvent(
                id: makeId(),
                title: template.name,
                type: .workout,
                scheduledAt: cal.date(from: components)!,
                recurrenceType: .weekly,
                recurrenceDays: [weekday],
                templateId: template.id
            ))
        }
        return events
    }

    // MARK: - Meals

    private func buildMealEvents(
        _ mealTemplates: [MealTemplateRef],
        now: Date,
        makeId: () -> String
    ) -> [ScheduledEvent] {
        let cal = Calendar(identifier: .gregorian)
        let today = cal.startOfDay(for: now)
        let times = mealTimes()
        var events: [ScheduledEvent] = []

        let generated = mealTemplates.filter { $0.origin == .generated }
        let pool = generated.isEmpty ? mealTemplates : generated

        for (i, slot) in times.enumerated() {
            let template = pool.isEmpty ? nil : pool[i % pool.count]
            var c = cal.dateComponents([.year, .month, .day], from: today)
            c.hour = slot.hour; c.minute = slot.minute

            events.append(ScheduledEvent(
                id: makeId(),
                title: template?.name ?? slot.label,
                type: .meal,
                scheduledAt: cal.date(from: c)!,
                recurrenceType: .daily,
                templateId: template?.id
            ))
        }
        return events
    }

    // MARK: - Sleep

    private static let bedtimeHour = 22

    private func buildSleepEvent(now: Date, makeId: () -> String) -> ScheduledEvent {
        let cal = Calendar(identifier: .gregorian)
        let today = cal.startOfDay(for: now)
        var c = cal.dateComponents([.year, .month, .day], from: today)
        c.hour = Self.bedtimeHour

        return ScheduledEvent(
            id: makeId(),
            title: text("Sleep", "שינה"),
            type: .sleep,
            scheduledAt: cal.date(from: c)!,
            recurrenceType: .daily
        )
    }

    // MARK: - Helpers

    private struct MealSlot {
        let label: String
        let hour: Int
        let minute: Int
    }

    private func mealTimes() -> [MealSlot] {
        switch profile.mealCountPerDay {
        case "2":
            return [
                MealSlot(label: text("Lunch", "ארוחת צהריים"), hour: 12, minute: 30),
                MealSlot(label: text("Dinner", "ארוחת ערב"), hour: 19, minute: 0),
            ]
        case "4":
            return [
                MealSlot(label: text("Breakfast", "ארוחת בוקר"), hour: 8, minute: 0),
                MealSlot(label: text("Lunch", "ארוחת צהריים"), hour: 12, minute: 30),
                MealSlot(label: text("Snack", "חטיף"), hour: 16, minute: 0),
                MealSlot(label: text("Dinner", "ארוחת ערב"), hour: 19, minute: 30),
            ]
        case "intermittent_fasting_16_8":
            return [
                MealSlot(label: text("Lunch", "ארוחת צהריים"), hour: 12, minute: 0),
                MealSlot(label: text("Snack", "חטיף"), hour: 16, minute: 0),
                MealSlot(label: text("Dinner", "ארוחת ערב"), hour: 19, minute: 30),
            ]
        default:
            return [
                MealSlot(label: text("Breakfast", "ארוחת בוקר"), hour: 8, minute: 0),
                MealSlot(label: text("Lunch", "ארוחת צהריים"), hour: 12, minute: 30),
                MealSlot(label: text("Dinner", "ארוחת ערב"), hour: 19, minute: 0),
            ]
        }
    }

    static func nextWeekdayOnOrAfter(_ from: Date, _ weekday: Int, cal: Calendar = .init(identifier: .gregorian)) -> Date {
        let current = cal.component(.weekday, from: from)
        let delta = (weekday - current + 7) % 7
        return cal.date(byAdding: .day, value: delta, to: from)!
    }

    private static let weekdayFillOrder = [1, 2, 3, 4, 5, 6, 7]

    private func getScheduleDays() -> [Int] {
        let count = min(max(profile.trainingDaysPerWeek, 1), 7)
        switch count {
        case 1: return [1] // Sunday
        case 2: return [1, 4] // Sun, Wed
        case 3: return [1, 3, 5] // Sun, Tue, Thu
        case 4: return [1, 2, 4, 5] // Sun, Mon, Wed, Thu
        default: return Array(Self.weekdayFillOrder.prefix(count))
        }
    }
}
